### PR TITLE
fix(MFA Issuer React): MFA Issuer is not displaying

### PR DESCRIPTION
**Pull request by ErikCH**
#### Description of changes
The `totpIssuer` and `totpUsername` is not displaying after being updated

#### Issue #, if available
#3092 

#### Description of how you validated changes
Added test

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are updated
- [x] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated
- [x] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/e2e/cypress/integration/ui/components/authenticator/sign-in-totp-mfa/sign-in-totp-mfa.steps.ts [hunk 1]
```diff
@@ -1,4 +1,7 @@
 import { Then, When } from '@badeball/cypress-cucumber-preprocessor';
+import { Decoder } from '@nuintun/qrcode';
```
Comment (ErikCH) [inline packages/e2e/cypress/integration/ui/components/authenticator/sign-in-totp-mfa/sign-in-totp-mfa.steps.ts:2]: This library is a dev dependency used to get the QR code information directly from the image

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/types.ts [hunk 2]
```diff
@@ -38,6 +38,7 @@ export interface UseAuthenticator extends AuthenticatorServiceFacade {
   getTotpSecretCode: () => Promise<string>;
 
   /** @deprecated For internal use only */
+  QRFields: { totpIssuer?: string; totpUsername?: string };
```
Comment (ErikCH) [inline packages/react-core/src/Authenticator/hooks/useAuthenticator/types.ts:41]: I made this deprecated internal, because we eventually want to handle this differently.

File: packages/e2e/features/ui/components/authenticator/sign-in-totp-mfa.feature [hunk 3]
```diff
@@ -12,6 +12,8 @@ Feature: Sign In with TOTP MFA
     And I type my password
     And I click the "Sign in" button
     Then I will be redirected to the setup totp page
+    Then I check to see if QR code is correct
+    Then I see the "Confirm" button
```
Comment (ErikCH) [inline packages/e2e/features/ui/components/authenticator/sign-in-totp-mfa.feature:16]: This is well more suited to be tested in a Cypress e2e test then a unit test IMO

File: packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx [hunk 4]
```diff
@@ -41,9 +41,8 @@ export const SetupTOTP = ({
   const [qrCode, setQrCode] = React.useState<string>();
   const [copyTextLabel, setCopyTextLabel] = React.useState<string>('COPY');
   const [secretKey, setSecretKey] = React.useState<string>('');
-
   const { totpIssuer = 'AWSCognito', totpUsername = user?.username } =
-    (fields as LegacyQRFields)?.QR ?? {};
```
Comment (ErikCH) [inline packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx:46]: This isn't in the fields, since it gets deleted in the `formFields.ts` file. Otherwise it creates an unusual `QR` input field on the `SetupTOTP` page. Since the behavior of `QR` doesn't match up with our other override field logic, I created a new `QRFields` from the `useAuthenticator` hook, that retrieves this information to use.

Comment (ErikCH) [reply packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx:46]: It's worth noting Angular and Vue do not use this logic, and garb the field data out of the xState context directly. React Native doesn't seem to allow issuer overrides, so this should not affect it.

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/types.ts (hunk 2 above)
Comment (wlee221) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/types.ts:41]: Should `fields` also be `@deprecated?

Comment (ErikCH) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/types.ts:41]: OK, I'll update it!

Comment (ErikCH) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/types.ts:41]: [d681c00](https://github.com/aws-amplify/amplify-ui/pull/3183/commits/d681c00fe7843e52b7b95834c9092c64c308781d)

File: packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx (hunk 4 above)
Comment (wlee221) [reply packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx:46]: Curious, why doesn't current code work and whether it's a regression. If so, we should avoid adding an extra return value from `useAuthenticator`, and resolve that regression instead. 

Ideally, `fields` should be the source of truth. Is `fields` from `useAuthenticator` not getting updated properly?

File: examples/angular/src/pages/ui/components/authenticator/sign-in-totp-mfa/sign-in-totp-mfa.component.html [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<amplify-authenticator>
+<amplify-authenticator [formFields]="formFields">
```
Comment (reesscot) [inline examples/angular/src/pages/ui/components/authenticator/sign-in-totp-mfa/sign-in-totp-mfa.component.html:1]: Why is this necessary? Is this required for all customers using TOT{? Could it be considered a breaking change?

File: packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx [hunk 6]
```diff
@@ -25,7 +25,7 @@ export const SetupTOTP = ({
   className,
   variation,
 }: RouteProps): JSX.Element => {
-  const { fields, getTotpSecretCode, isPending, user } = useAuthenticator(
+  const { getTotpSecretCode, isPending, user, QRFields } = useAuthenticator(
```
Comment (reesscot) [inline packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx:28]: Is this a new external facing API, or is this internal only?

**Review by reesscot — COMMENTED** [review]
Left some questions

File: packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx (hunk 6 above)
Comment (ErikCH) [reply packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx:28]: This is internal only.

File: examples/angular/src/pages/ui/components/authenticator/sign-in-totp-mfa/sign-in-totp-mfa.component.html (hunk 5 above)
Comment (ErikCH) [reply examples/angular/src/pages/ui/components/authenticator/sign-in-totp-mfa/sign-in-totp-mfa.component.html:1]: This is the examples app. I added a new e2e test that now checks the setup TOTP QR field. So I had to add this new fields in the vue/angular/react e2e tests.

File: packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx (hunk 4 above)
Comment (ErikCH) [reply packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx:46]: In the original implementation Vue, React and Angular all looked up the issuer information directly from the actor context in the SetupTOTP file. 

Even though the `QR` object is passed into the `formFields` prop it really has never been used as as form field, since it really isn't one. In fact it's deleted out of the form fields as seen below.

```
export const getFormFields = (
  route: FormFieldComponents,
  state: AuthMachineState
): FormFields => {
...
  delete formFields['QR'];
  return formFields;
```
When the big `react-core` refactor was done, it was assumed it was in the fields, but this was never the case. Vue and React still pulled it directly from the context.

If we keep it in the fields, then we'll need to update the `FormFields` components for each framework and ignore the `QR` field, since it doesn't exist. The `formFields` prop is used to help generate all form fields, even those that don't exist. 

I'm up for suggestions on a better way of handling this. I noticed in the refactor a `getTotpSecretCode` was added, so I assumed it was ok to add another TOTP/QR return value.

Comment (ErikCH) [reply packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx:46]: @calebpollman Well QR isn't a formField, it's an image, but we kept all the logic for updating and customizing components, like dialCode, etc, in `formField`. In our next breaking change we can remove it from there.

File: packages/e2e/package.json [hunk 7]
```diff
@@ -24,6 +24,7 @@
     "@aws-amplify/ui": "^5.3.1",
     "@badeball/cypress-cucumber-preprocessor": "^12.1.0",
     "@bahmutov/cypress-esbuild-preprocessor": "^2.1.3",
+    "@nuintun/qrcode": "^3.3.0",
```
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/e2e/package.json:27]: @calebpollman I believe a cypress test, testing the QR code generated, and looking at the image is the best way of doing this. And since it's a dev dependency, for the Cypress e2e test, the impact is minimal.

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/__tests__/utils.spec.tsx [hunk 8]
```diff
@@ -13,9 +13,27 @@ import {
   getComparator,
   getMachineFields,
   getTotpSecretCodeCallback,
+  getQRFields,
 } from '../utils';
+import { AuthActorContext } from '@aws-amplify/ui';
```
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/__tests__/utils.spec.tsx:18]: I'll update it.

File: packages/e2e/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (calebpollman) [reply packages/e2e/package.json:27]: While i appreciate that you feel that it is the best way, is there any other way to test the output of the QR code without adding a 3p dep?

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/__tests__/utils.spec.tsx [hunk 9]
```diff
@@ -145,4 +163,23 @@ describe('getMachineFields', () => {
...
   });
+
+  describe('getQRFields', () => {
+    const state = {} as unknown as AuthMachineState;
+
+    beforeEach(() => {
+      jest.clearAllMocks();
+    });
+
+    it('returns the correct QR issuers', () => {
```
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/__tests__/utils.spec.tsx:174]: I'll change to issuer and username.

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/useAuthenticator.tsx [hunk 10]
```diff
@@ -43,6 +44,12 @@ export default function useAuthenticator(
   // `fields` from updating with current form state on value changes
   const serviceSnapshot = service.getSnapshot();
 
+  // legacy `QRFields` values only used for SetupTOTP page to retrieve issuer information, will be removed in future
+  const QRFields = useMemo(
+    () => getQRFields(serviceSnapshot as AuthMachineState),
+    [serviceSnapshot]
+  );
```
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/useAuthenticator.tsx:51]: Unfortunately, I don't really see an easy way of keeping it in the existing fields variable. We'd have to explicitly not delete it, then have to make changes to Vue, React and Angular so the Form Fields components won't automatically try to create a new bogus form field component. On the other hand, if we keep it in `useAuthenticator` it will only affect React, and won't require anything additional.

I'm up for more suggestions if you know of an easier way. Otherwise, we could wait until the next breaking change to fix this.

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/utils.ts [hunk 11]
```diff
@@ -52,6 +53,16 @@ export const getComparator =
     return areSelectorDepsEqual(currentSelectorDeps, nextSelectorDeps);
   };
 
+export const getQRFields = (
+  state: AuthMachineState
+): { totpIssuer?: string; totpUsername?: string } => {
+  const fields = getActorContext(state);
```
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/utils.ts:59]: Well technically it's in actorContext, and it's all fields.

File: packages/e2e/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/e2e/package.json:27]: Not that I know of.

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/__tests__/utils.spec.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/__tests__/utils.spec.tsx]: [0bfdc84](https://github.com/aws-amplify/amplify-ui/pull/3183/commits/0bfdc841fa6e83628660407cec2848c7b6cf1360)

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/__tests__/utils.spec.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/__tests__/utils.spec.tsx]: [0bfdc84](https://github.com/aws-amplify/amplify-ui/pull/3183/commits/0bfdc841fa6e83628660407cec2848c7b6cf1360)

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/useAuthenticator.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (calebpollman) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/useAuthenticator.tsx:51]: Why can't we just put the workaround in `SetupTOTP`?

File: packages/e2e/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/e2e/package.json:27]: Our existing QRCode library does not decode, -> https://github.com/soldair/node-qrcode/issues/129

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/useAuthenticator.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/useAuthenticator.tsx:51]: @calebpollman The problem is how we get it into the `SetupTOTP` component? It's not in the `fields` component passed from `useAuthenticator`. That's why it doesn't update the QR image for the customer.

File: packages/react/src/components/Authenticator/shared/FormFields.tsx [hunk 12]
```diff
@@ -5,9 +5,10 @@ import { useAuthenticator } from '@aws-amplify/ui-react-core';
 
 export function FormFields(): JSX.Element {
   const { fields } = useAuthenticator(({ route }) => [route]);
+  const removedQRFields = fields.filter((field) => field.name !== 'QR');
```
Comment (ErikCH) [inline packages/react/src/components/Authenticator/shared/FormFields.tsx:8]: Since we add in QR fields, we need to remove it back out again, so another field isn't created.

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/useAuthenticator.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/useAuthenticator.tsx]: Per offline discussion, I have removed the `QRFields` from `useAuthenticator`. We are now grabbing the information in `getMachineFields` instead, and then deleting it in `FormFields` component, since it's not technically a field that needs to be displayed.

File: packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx (hunk 4 above)
Comment (ErikCH) [reply packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx:46]: Return value from `useAuthenticator` has been removed.

File: packages/react/src/components/Authenticator/shared/FormFields.tsx (hunk 12 above)
Comment (ErikCH) [reply packages/react/src/components/Authenticator/shared/FormFields.tsx]: reverted

File: packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx (hunk 4 above)
Comment (ErikCH) [reply packages/react/src/components/Authenticator/SetupTOTP/SetupTOTP.tsx:46]: Reverted back, after offline discussion with @calebpollman - Since it affects React native. `useAuthenticator` is now updated with new `QRFields`

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/useAuthenticator.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (ErikCH) [reply packages/react-core/src/Authenticator/hooks/useAuthenticator/useAuthenticator.tsx:51]: Reverted - added back in `QRFields` after offline discussion with @calebpollman

**Review by calebpollman — APPROVED** [review]
@ErikCH Thanks for sorting through this! Appreciate the work 🚢

**Review by wlee221 — APPROVED** [review]
